### PR TITLE
New: Support for Destructuring

### PR DESCRIPTION
**Pull request by xjamundx**
- [x] Basic Destructuring Assignment
- [x] Nested Desctructured Assignment
- [x] Destructured Params
- [x] Destructured Params with Defaults
- [x] Destructured Params (Arrow Functions)
- [x] Destructured Params with Defaults (Arrow Functions)
- [ ]  Destructured Params with Defaults (Shorthand Syntax)

Tracking the functionality mostly by looking at the 13 related PRs here:
https://github.com/jquery/esprima/pulls?q=destructuring

It seems like unlike most ES6 features destructuring hasn't landed in Esprima 2.0 yet, so I'll continue to work off the `harmony` version.

Grabbing the tests from here:
https://github.com/jquery/esprima/blob/harmony/test/harmonytest.js

### CODE REVIEW

**Comment by xjamundx** [comment]
All the fixture tests pass, but something else is broken in the secondary tests. Definitely out of time to work on this for the week....PR is a bit messy, but could probably use a review. I think we're very close.

Code is still a bit messy.....

File: espree.js [hunk 1]
```diff
@@ -2550,7 +2571,7 @@ function parseObjectProperty() {
          * If shorthand properties aren't allow, then this is an automatic
          * syntax error.
          */
-        if (computed || !allowShorthand) {
+        if (computed && !(allowShorthand)) {
```
Comment (xjamundx) [inline espree.js]: not needed

File: espree.js [hunk 2]
```diff
@@ -2489,7 +2510,7 @@ function parseObjectProperty() {
             );
         }
 
-        if (token.value === "set" && !match(":") && !match("(")) {
+        if (token.value === "set" && !(match(":") || match("("))) {
```
Comment (xjamundx) [inline espree.js]: probably not needed

File: espree.js [hunk 3]
```diff
@@ -2471,7 +2492,7 @@ function parseObjectProperty() {
          * Check for getters and setters. Be careful! "get" and "set" are legal
          * method names. It's only a getter or setter if followed by a space.
          */
-        if (token.value === "get" && !match(":") && !match("(")) {
+        if (token.value === "get" && !(match(":") || match("("))) {
```
Comment (xjamundx) [inline espree.js]: probably not needed (though new way is closer to current `harmony`

File: espree.js [hunk 4]
```diff
@@ -264,6 +257,34 @@ function skipComment() {
     }
 }
 
+// fancy way to write a map
```
Comment (xjamundx) [inline espree.js]: def need to put in its own file

**Comment by nzakas** [comment]
Be on the lookout for how the spread operator pops up in destructuring.

File: espree.js [hunk 5]
```diff
@@ -2595,49 +2616,56 @@ function parseObjectProperty() {
             )
         );
 
-    } else {
```
Comment (xjamundx) [inline espree.js]: oh yeah this used to be an `else` the current `harmony` prefers the early return style.  changed it during some debugging...

File: espree.js [hunk 6]
```diff
@@ -2646,40 +2674,31 @@ function parseObjectInitialiser() {
...
-            key = "$" + name;
-            if (Object.prototype.hasOwnProperty.call(kindMap, key)) {
-                if (kindMap[key] === PropertyKind.Data) {
+            if (kindMap.has(name)) {
+                storedKind = kindMap.get(name);
+                if (storedKind === PropertyKind.Data) {
+                    // FYI this first one is not in ESPRIMA I think
                     if (kind === PropertyKind.Data && name === "__proto__" && allowDuplicates) {
-                        // Duplicate '__proto__' literal properties are forbidden in ES 6
+                        // Duplicate "__proto__' literal properties are forbidden in ES 6
```
Comment (xjamundx) [inline espree.js]: annoyingly changed this by accident

File: espree.js [hunk 7]
```diff
@@ -3412,18 +3519,29 @@ function parseVariableIdentifier() {
 }
 
 function parseVariableDeclaration(kind) {
-    var init = null, id,
-        marker = markerCreate();
-
-    id = parseVariableIdentifier();
-
-    // 12.2.1
-    if (strict && syntax.isRestrictedWord(id.name)) {
```
Comment (xjamundx) [inline espree.js]: need to check what this was doing. replaced with current `harmony`, but it may have had some value...

**Comment by xjamundx** [comment]
Finally all of the tests pass. I may be missing some edge case and negative testing, but I'm pretty stoked.

**Comment by nzakas** [comment]
Cool! Can you squash into one commit?

File: tests/fixtures/ecma-features-mix/destructuring-and-defaultParams/param-object-short.config.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+	defaultParams: true,
+	destructuring: true,
+    objectLiteralShorthandMethods: true
```
Comment (xjamundx) [inline tests/fixtures/ecma-features-mix/destructuring-and-defaultParams/param-object-short.config.js]: spacing

**Comment by xjamundx** [comment]
So everything works except for these:
- `function a({x = 10}) {}`
- `function a({x = 10, y: { z = 10 }}) {}`

Both of these fail in `esprima` and in `esprima-fb`, so I'm unsure how to proceed :)

See also https://github.com/jquery/esprima/issues/1005

**Comment by nzakas** [comment]
Can you for an issue with those cases? We can handle them as a bug and then merge this.

**Comment by xjamundx** [comment]
done!
